logging/nxscope/nxscope_chan.c

```diff
@@ -0,0 +1,1566 @@
+{
+  size_t reqlen    = 0;
+  size_t type_size = 0;
+  int    ret       = OK;
+
+  DEBUGASSERT(s);
+
+  /* We use one additiona buffer to handle no-buffered critical channels.
+   * We have to allocate memory to fit at least single frame with
+   * the longest possible critical channel sample data.
+   */
+
+  type_size = g_type_size[dtype];
+  reqlen = (s->proto_stream->hdrlen + 1 + type_size * vdim +
+            mlen + s->proto_stream->footlen);
+
+  /* This is the first time allocation or we need more memory */
+
+  if (reqlen > s->cri_len)
+    {
+      s->cribuf = realloc(s->cribuf, reqlen);
+      if (s->cribuf == NULL)
+        {
+          ret = -errno;
+          _err("ERROR: cribuf realloc failed %d\n", ret);
+          goto errout;
+        }
+
+      s->cri_len = reqlen;
+    }
+
+errout:
+  return ret;
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxscope_chan_init
+ *
+ * Description:
+ *   Initialize nxscope channel
+ *
+ * Input Parameters:
+ *   s    - a pointer to a nxscope instance
+ *   ch   - a channel id
+ *   name - a channel name
+ *   type - a channel data type (union nxscope_chinfo_type_u)
+ *   vdim - a vector data dimension (vdim=1 for a point)
+ *   mlen - a length of metadata
+ *
+ ****************************************************************************/
+
+int nxscope_chan_init(FAR struct nxscope_s *s, uint8_t ch, FAR char *name,
+                      uint8_t type, uint8_t vdim, uint8_t mlen)
+{
+  union nxscope_chinfo_type_u utype;
+  size_t                      txbuf_len = 0;
+  int                         ret = OK;
+
+  DEBUGASSERT(s);
+  DEBUGASSERT(name);
+
+  if (ch > s->cmninfo.chmax)
+    {
+      _err("ERROR: invalid channel %d\n", ch);
+      ret = -EINVAL;
+      goto errout;
+    }
+
+  nxscope_lock(s);
+
+  utype.u8 = type;
+  if (utype.s.cri)
+#ifdef CONFIG_LOGGING_NXSCOPE_CRICHANNELS
+    {
+      /* Allocate or re-allocate critical channels buffer */
+
+      ret = nxscope_cribuf_alloc(s, utype.s.dtype, vdim, mlen);
+      if (ret < 0)
+        {
+          _err("ERROR: nxscope_cribuf_alloc failed %d\n", ret);
+          ret = -EINVAL;
+          goto errout;
+        }
+    }
+#else
+    {
+      _err("ERROR: cri channels not supported ch=%d\n", ch);
+      ret = -EINVAL;
+      goto errout;
+    }
+#endif
+
+  /* Reallocate TX buffer if needed
+   *
+   * REVISIT:
+   *   In the worst case this give us chmax reallocationsm but we don't know
+   *   how much memory is needed for the channel with the longest channel
+   *   name. Or just allocate memor for CHAN_NAMELEN_MAX ?
+   */
+
+  txbuf_len = (sizeof(struct nxscope_chinfo_s) +
+               strnlen(name, CHAN_NAMELEN_MAX) + 1);
+  ret = nxscope_txbuf_realloc(s, txbuf_len);
```

Comment:
Same as with zalloc/free. realloc is free to call `free()`, this in turn will trash your memory eventually (but never during tests, only on production :)). Either allocate only once `CHAN_NAMELEN_MAX` on heap, or (better yet) keep this on users stack.

Reply:
After second thought I agree. This realloc doesn't make too much sense. With realloc we can save up to CHAN_NAMELEN_MAX memory bytes, which is insignificant.  I'll fix it.